contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,128 @@
+ * spoke pool. It rejects messages which do not originate from the cross domain admin, which should be set to the hub pool.
+ * @dev Base contract designed to be deployed on L2 to re-route messages from L1 to L3. If a message is sent to this contract which
+ * came from the L1 cross domain admin, then it is routed to the L3 via the canonical messaging bridge. Tokens sent from L1 are stored
+ * temporarily on this contract. Any EOA can initiate a bridge of these tokens to the target `l3SpokePool`.
+ * @custom:security-contact bugs@across.to
+ */
+abstract contract ForwarderBase is UUPSUpgradeable {
+    // L3 address of the recipient of L1 messages and tokens sent via the re-router adapter on L1.
+    address public l3SpokePool;
+
+    // L1 address of the contract which can relay messages to the l3SpokePool contract and update this proxy contract.
+    address public crossDomainAdmin;
+
+    event TokensForwarded(address indexed l2Token, uint256 amount);
+    event MessageForwarded(address indexed target, bytes message);
+    event SetXDomainAdmin(address indexed crossDomainAdmin);
+    event SetL3SpokePool(address indexed l3SpokePool);
+
+    error InvalidCrossDomainAdmin();
+    error InvalidL3SpokePool();
+
+    /*
+     * @dev All functions with this modifier must revert if msg.sender != crossDomainAdmin. Each L2 may have
+     * unique aliasing logic, so it is up to the chain-specific forwarder contract to verify that the sender is valid.
+     */
+    modifier onlyAdmin() {
+        _requireAdminSender();
+        _;
+    }
+
+    /**
+     @notice Constructs the Forwarder contract.
+     @dev _disableInitializers() restricts anybody from initializing the implementation contract, which if not done, 
+     * may disrupt the proxy if another EOA were to initialize it.
+     */
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializes the forwarder contract.
+     * @param _l3SpokePool L3 address of the contract which will receive messages and tokens which are temporarily
+     * stored in this contract on L2.
+     * @param _crossDomainAdmin L1 address of the contract which can send root bundles/messages to this forwarder contract.
+     * In practice, this is the hub pool.
+     */
+    function __Forwarder_init(address _l3SpokePool, address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
+        _setL3SpokePool(_l3SpokePool);
+        _setCrossDomainAdmin(_crossDomainAdmin);
+    }
+
+    // Added so that this contract may receive ETH to fund transaction gas fees.
+    receive() external payable {}
+
+    /**
+     * @dev When called by the cross domain admin (i.e. the hub pool), the msg.data should be some function
+     * recognizable by the L3 spoke pool, such as "relayRootBundle" or "upgradeTo". Therefore, we forward
+     * this message to the L3 spoke pool using the implemented messaging logic of the L2 forwarder.
```

Comment:
Is this always the case? Doesn't the hub pool sometimes intend to call this contract, not the L3 contract? For example, the rerouter adapter will call relayTokens/relayMessage() on this contract, right?

Also, how does the hubpool call `setCrossDomainAdmin` on the L3 spoke pool? In this case it wouldn't hit the `fallback` because this function is implemented here, right?

Reply:
That's very true. I removed these functions here https://github.com/across-protocol/contracts/pull/609/commits/19813c6c0a7f304969f64249b9a8ea213c804815. Now all messages destined for L3 should be called by `relayMessage` (although this will be reflected in this PR https://github.com/across-protocol/contracts/pull/610).